### PR TITLE
[node-manager](caps) Add the ability to disable StaticInstance bootstrapping

### LINKED ISSUE #7370: [node-manager] It is very difficult to delete a StaticInstance with a non-Pending status
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

It is very difficult to delete a StaticInstance custom resource with a non-Pending status. If we try to delete a StaticInstance with a status other than Pending, we will get an error from the validation webhook because it is a forbidden operation. That's why we need to change the status of the StaticInstance to Pending before we can delete it. By convention we need to find and delete the corresponding Instance resource, this solution has some disadvantages:
1. Race conditions
1. It's not user-friendly

We need to provide a simple solution to delete a StaticInstance custom resource.

### Proposed Solution

In the `StaticMachineTemplate` custom resource we need to use the Helm merge function to merge the labelSelector with the map below:
```yaml
matchExpressions: 
  key: allow-bootstrap
  operator: NotIn
  values: 
    - "false"
```

If after this change we add the label `allow-bootstrap: "false"` to a StaticInstance, then the StaticInstance will leave the NodeGroup and the status will change to `Pending` and we can delete the StaticInstance.
We already have functionality in the StaticInstance controller that checks if the StaticInstance has left the NodeGroup and changes the status to `Pending`.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by artbin**
## Description

Add the `"node.deckhouse.io/allow-bootstrap": "false"` label requirement to the `StaticInstance` label selector. Add a detailed description of `StaticInstance` deletion to the documentation.

## Why do we need it, and what problem does it solve?

Close #7370.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: node-manager
type: fix
summary: Add the ability to disable StaticInstance bootstrapping by adding the '"node.deckhouse.io/allow-bootstrap": "false"' label.
impact_level: default
```

### CODE REVIEW

**Review by name212 — APPROVED** [review]

File: modules/040-node-manager/docs/FAQ.md [hunk 1]
```diff
@@ -129,7 +129,10 @@ bash /var/lib/bashible/cleanup_static_node.sh --yes-i-am-sane-and-i-understand-w
 
 A `StaticInstance` that is in the `Pending` state can be deleted with no adverse effects.
 
-To delete a `StaticInstance` in any state other than `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), you need to delete the corresponding  [Instance](cr.html#instance) resource, and then the `StaticInstance` will be deleted automatically.
+To delete a `StaticInstance` in any state other than `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), you need to:
```
Comment (RomanenkoDenys) [inline modules/040-node-manager/docs/FAQ.md:132]:
```suggestion
To delete a `StaticInstance` in any state other than `Pending` (`Running`, `Cleaning`, `Bootstraping`), you need to:
```

File: modules/040-node-manager/docs/FAQ_RU.md [hunk 2]
```diff
@@ -132,7 +132,10 @@ bash /var/lib/bashible/cleanup_static_node.sh --yes-i-am-sane-and-i-understand-w
 
 `StaticInstance`, находящийся в состоянии `Pending` можно удалять без каких-либо проблем.
 
-Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), нужно удалить соответствующий ресурс [Instance](cr.html#instance), после чего `StaticInstance` удалится автоматически.
+Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Runnig`, `Cleaning`, `Bootstraping`):
```
Comment (RomanenkoDenys) [inline modules/040-node-manager/docs/FAQ_RU.md:135]:
```suggestion
Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Running`, `Cleaning`, `Bootstraping`):
```

**Review by RomanenkoDenys — COMMENTED** [review]
some errors

File: modules/040-node-manager/docs/FAQ.md [hunk 3]
```diff
@@ -129,7 +129,10 @@ bash /var/lib/bashible/cleanup_static_node.sh --yes-i-am-sane-and-i-understand-w
 
 A `StaticInstance` that is in the `Pending` state can be deleted with no adverse effects.
 
-To delete a `StaticInstance` in any state other than `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), you need to delete the corresponding  [Instance](cr.html#instance) resource, and then the `StaticInstance` will be deleted automatically.
+To delete a `StaticInstance` in any state other than `Pending` (`Running`, `Cleaning`, `Bootstraping`), you need to:
```
Comment (z9r5) [inline modules/040-node-manager/docs/FAQ.md:132]:
```suggestion
To delete a `StaticInstance` in any state other than `Pending` (`Running`, `Cleaning`, `Bootstrapping`), you need to:
```

File: modules/040-node-manager/docs/FAQ_RU.md [hunk 4]
```diff
@@ -132,7 +132,10 @@ bash /var/lib/bashible/cleanup_static_node.sh --yes-i-am-sane-and-i-understand-w
 
 `StaticInstance`, находящийся в состоянии `Pending` можно удалять без каких-либо проблем.
 
-Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), нужно удалить соответствующий ресурс [Instance](cr.html#instance), после чего `StaticInstance` удалится автоматически.
+Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Running`, `Cleaning`, `Bootstraping`):
```
Comment (z9r5) [inline modules/040-node-manager/docs/FAQ_RU.md:135]:
```suggestion
Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Running`, `Cleaning`, `Bootstrapping`):
```

**Review by z9r5 — APPROVED** [review]

**Comment by artbin** [comment]
Manual testing
=============

Add `allow-bootstrap: "false"` label:

Screenshot 2024-02-14 at 14 24 44


Remove `allow-bootstrap: "false"` label:

Screenshot 2024-02-14 at 14 25 06

**Comment by RomanenkoDenys** [comment]
@artbin, one note from the nabokihms that the label should be in the deckhouse.io group.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @artbin. I've started the workflow for label 'e2e/run/static'...

**Review by nabokihms — APPROVED** [review]

File: modules/040-node-manager/images/caps-controller-manager/src/internal/scope/machine_scope.go [hunk 5]
```diff
@@ -30,6 +31,17 @@ import (
 	infrav1 "caps-controller-manager/api/infrastructure/v1alpha1"
 )
 
+var allowBootstrapRequirement *labels.Requirement
+
+func init() {
+	var err error
+
+	allowBootstrapRequirement, err = labels.NewRequirement("deckhouse.io/allow-bootstrap", selection.NotIn, []string{"false"})
```
Comment (nabokihms) [inline modules/040-node-manager/images/caps-controller-manager/src/internal/scope/machine_scope.go:39]:
```suggestion
	allowBootstrapRequirement, err = labels.NewRequirement("node.deckhouse.io/allow-bootstrap", selection.NotIn, []string{"false"})
```

File: modules/040-node-manager/images/caps-controller-manager/src/internal/scope/machine_scope.go [hunk 6]
```diff
@@ -122,15 +134,15 @@ func (m *MachineScope) HasFailed() bool {
...
 	}
 
 	labelSelector, err := metav1.LabelSelectorAsSelector(m.StaticMachine.Spec.LabelSelector)
 	if err != nil {
 		return nil, errors.Wrap(err, "unable to convert StaticMachine label selector")
 	}
 
-	return labelSelector, nil
+	return labelSelector.Add(*allowBootstrapRequirement), nil
 }
```
Comment (nabokihms) [inline modules/040-node-manager/images/caps-controller-manager/src/internal/scope/machine_scope.go:146]: Do we need a unit test for this? Probably no, but it is better to ask 🙂

**Review by nabokihms — CHANGES_REQUESTED** [review]
There is only one suggestion left. Overall, LGTM.

File: modules/040-node-manager/docs/FAQ.md [hunk 7]
```diff
@@ -129,7 +129,11 @@ bash /var/lib/bashible/cleanup_static_node.sh --yes-i-am-sane-and-i-understand-w
 
 A `StaticInstance` that is in the `Pending` state can be deleted with no adverse effects.
 
-To delete a `StaticInstance` in any state other than `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), you need to delete the corresponding  [Instance](cr.html#instance) resource, and then the `StaticInstance` will be deleted automatically.
+To delete a `StaticInstance` in any state other than `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), you need to:
+1. Add the label `"deckhouse.io/allow-bootstrap": "false"` to the `StaticInstance`.
```
[the remark this answers is not in the document]
Comment (artbin) [reply modules/040-node-manager/docs/FAQ.md]: done

File: modules/040-node-manager/docs/FAQ_RU.md [hunk 8]
```diff
@@ -132,7 +132,11 @@ bash /var/lib/bashible/cleanup_static_node.sh --yes-i-am-sane-and-i-understand-w
 
 `StaticInstance`, находящийся в состоянии `Pending` можно удалять без каких-либо проблем.
 
-Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Runnig`, `Cleaning`, `Bootstraping`), нужно удалить соответствующий ресурс [Instance](cr.html#instance), после чего `StaticInstance` удалится автоматически.
+Чтобы удалить `StaticInstance` находящийся в любом состоянии отличном от `Pending` (`Runnig`, `Cleaning`, `Bootstraping`):
+1. Добавьте лейбл `"deckhouse.io/allow-bootstrap": "false"` в `StaticInstance`.
```
[the remark this answers is not in the document]
Comment (artbin) [reply modules/040-node-manager/docs/FAQ_RU.md]: done

File: modules/040-node-manager/images/caps-controller-manager/src/internal/scope/machine_scope.go (hunk 5 above)
Comment (artbin) [reply modules/040-node-manager/images/caps-controller-manager/src/internal/scope/machine_scope.go]: done

**Review by nabokihms — APPROVED** [review]

**Pull request merged**